tools/PI/DevHome.PI/PIApp.xaml.cs

```diff
@@ -48,6 +49,9 @@ public App()
                 services.AddSingleton<PIAppInfoService>();
                 services.AddSingleton<PIInsightsService>();
                 services.AddSingleton<WERHelper>();
+                services.AddSingleton<WERAnalyzer>();

```

Comment:
Nit: For testing and mocking services, consider using interfaces.

Reply:
I'm guessing you can't really take advantage of dependency injection unless you're using interfaces, correct? Definitely something we should start moving towards.

Reply:
I am not aware of a strict rule/best practice here, it more depends on how we intend to use a class. The way I approach it is by asking questions: Is this code easily testable/reusable/scalable?

From my observation, here a pattern I've established:
- Services that we are not interested in testing or are impossible to test (for example: Service that installs something on your computer (depends on agent) or does some https calls (depends on network), etc...).
  - Those services ideally need to be mapped to an interface so that they can be mocked.
  - Example: Methods in this class need to be mockable. https://github.com/microsoft/devhome/blob/ac1e5978bc7491a598972be01214de1acf33487d/services/DevHome.Services.WindowsPackageManager/Services/WinGet.cs#L16
- Services that we want to test (for example, view models that call [mockable] services)
  - Those services are not required to have an interface, especially that they are unlikely to be injected into other services (there are exception to this, especially in cases where the view class is a registered type, and the view model is injected into the view class constructor).
  - Example: https://github.com/microsoft/devhome/blob/ac1e5978bc7491a598972be01214de1acf33487d/tools/SetupFlow/DevHome.SetupFlow.UnitTest/ViewModels/SearchViewModelTest.cs#L65-L71
- There are other kind of services as well such as the ones that are mapped to the same type for enumeration reasons. For example, `Add<IStorage, DiskStorage>()` `Add<IStorage, DBStorage>()`, etc.... As well as factory classes. Those need to be mapped to a parent type/interace.

Reply:
Good discussion. Definitely things to keep in mind when we get serious about testing this product.